docs/5.x/writing-a-record-builder.md

```diff
@@ -0,0 +1,250 @@
+
+Once you know what queries the `RecordBuilder`s you are going to create will execute, you can start coding.
+The first thing to do is implement the `getRecordMetadata()` method.
+
+All this method does is return a list of `Record` entries describing the records the builder will create:
+
+```
+use Piwik\ArchiveProcessor\Record;
+
+public function getRecordMetadata(ArchiveProcessor $archiveProcessor): array
+{
+    return [
+        Record::make(Record::TYPE_BLOB, 'MyPlugin_myRecord'),
+        Record::make(Record::TYPE_NUMERIC, 'MyPlugin_myMetric'),
+        ...
+    ];
+}
+```
+
+The above is a typical example of how this method would be implemented, but it doesn't have to just be a hard-coded array.
+You can use the `ArchiveProcessor` to get the current site/period/segment or fetch system settings or measurable
+settings and vary the result based on that information. The only requirement is that every `Record` returned matches
+what can be returned by the `aggregate()` method, which we'll look at next.
+
+### Step three: implement `aggregate()`
+
+The next step is to implement your actual log aggregation logic in the `aggregate()` method. This method accepts
+an `ArchiveProcessor` and returns an array mapping record names with record values to insert. Record values are
+either numeric metric values or `DataTable` instances, which get serialized and inserted as blobs.
+
+As for how they are created, well, there is no straightforward way to define how log aggregation is done.
+
+The current pattern in Matomo is to use the core `LogAggregator` class to query log data and loop through the result.
+If your plugin provides its own additional log tables, then the pattern is to define your own `Aggregator` classes
+to build and execute log aggregation SQL queries, and use those classes in your `RecordBuilders`.
+
+An example of this might look like:
+
+```
+public function aggregate(ArchiveProcessor $archiveProcessor): array
+{
+    $logAggregator = $archiveProcessor->getLogAggregator();
+    
+    $report = new DataTable();
+    
+    $query = $logAggregator->queryVisitsByDimension(['label' => 'config_browser_name']);
+    while ($row = $query->fetch()) {
+        $columns = [
+            Metrics::INDEX_NB_UNIQ_VISITORS => $row[Metrics::INDEX_NB_UNIQ_VISITORS],
+            Metrics::INDEX_NB_VISITS => $row[Metrics::INDEX_NB_VISITS],
+            Metrics::INDEX_NB_ACTIONS => $row[Metrics::INDEX_NB_ACTIONS],
+            Metrics::INDEX_NB_USERS => $row[Metrics::INDEX_NB_USERS],
+            Metrics::INDEX_MAX_ACTIONS => $row[Metrics::INDEX_MAX_ACTIONS],
+            Metrics::INDEX_SUM_VISIT_LENGTH => $row[Metrics::INDEX_SUM_VISIT_LENGTH],
+            Metrics::INDEX_BOUNCE_COUNT => $row[Metrics::INDEX_BOUNCE_COUNT],
+            Metrics::INDEX_NB_VISITS_CONVERTED => $row[Metrics::INDEX_NB_VISITS_CONVERTED],
+        ];
+
+        $report->sumRowWithLabel($row['label'] ?? '', $columns);
+    }
+
+    return [
+        'MyPlugin_myRecord' => $report,
```

Comment:
As we return `Record::make(Record::TYPE_NUMERIC, 'MyPlugin_myMetric'),` from the `getRecordMetadata` example above, should we also, for consistency, have it here with some sample metric value?

Reply:
thanks!